lib/src/rules/dont_compare_unrelated_types_for_equality.dart

```diff
@@ -0,0 +1,195 @@
+
+  if (other == instance) print('someFunction13'); // LINT
+}
+
+class ClassBase {}
+
+class DerivedClass1 extends ClassBase {}
+
+abstract class Mixin {}
+
+class DerivedClass2 extends ClassBase with Mixin {}
+
+class DerivedClass3 extends ClassBase implements Mixin {}
+```
+
+**GOOD:**
+```
+void someFunction2() {
+  var x = '1';
+  var y = '2';
+  if (x == y) print(someFunction2); // OK
+}
+```
+
+**GOOD:**
+```
+void someFunction3() {
+  for (var i = 0; i < 10; i++) {
+    if (i == 0) print(someFunction3); // OK
+  }
+}
+```
+
+**GOOD:**
+```
+void someFunction4() {
+  var x = '1';
+  if (x == null) print(someFunction4); // OK
+}
+```
+
+**GOOD:**
+```
+void someFunction7() {
+  List someList;
+
+  if (someList.length == 0) print('someFunction7'); // OK
+}
+```
+
+**GOOD:**
+```
+void someFunction8(ClassBase instance) {
+  DerivedClass1 other;
+
+  if (other == instance) print('someFunction8'); // OK
+}
+```
+
+**GOOD:**
+```
+void someFunction10(unknown) {
+  var what = unknown - 1;
+  for (var index = 0; index < unknown; index++) {
+    if (what == index) print('someFunction10'); // OK
+  }
+}
+```
+
+**GOOD:**
+```
+void someFunction11(Mixin instance) {
+  var other = new DerivedClass2();
+
+  if (other == instance) print('someFunction11'); // OK
+}
+
+class ClassBase {}
+
+abstract class Mixin {}
+
+class DerivedClass2 extends ClassBase with Mixin {}
+```
+''';
+
+class DontCompareUnrelatedTypesForEquality extends LintRule {
+  _Visitor _visitor;
+
+  DontCompareUnrelatedTypesForEquality() : super(
+      name: 'dont_compare_unrelated_types_for_equality',
+      description: _desc,
+      details: _details,
+      group: Group.errors,
+      maturity: Maturity.experimental) {
+    _visitor = new _Visitor(this);
+  }
+
+  @override
+  AstVisitor getVisitor() => _visitor;
+}
+
+class _Visitor extends SimpleAstVisitor {
+  static const _dynamicClassName = 'dynamic';
+
+  final LintRule rule;
+
+  _Visitor(this.rule);
+
+  @override
+  void visitBinaryExpression(BinaryExpression node) {
+    // TODO: This comparison is brittle, hopefully there will be only one bool.
+    if (node.parent is! IfStatement || node.staticType.name != 'bool' ||
+        node.operator.type != TokenType.EQ_EQ) {
+      return;
+    }
+
+    if (_differentTypes(node)) {
+      rule.reportLint(node);
+    }
+  }
+
+  bool _compareNonNull(DartType left, DartType right) =>
+      left != null && !left.isMoreSpecificThan(right);
+
+  bool _differentDynamicStatic(DartType type, Expression expression) =>
+      type.name == _dynamicClassName && expression is! NullLiteral;
+
+  bool _differentTypes(BinaryExpression node) {
+    final DartType leftStaticType = node.leftOperand.staticType;
+    final DartType rightStaticType = node.rightOperand.staticType;
+    final DartType leftPropagatedType = node.leftOperand.propagatedType;
+    final DartType rightPropagatedType = node.rightOperand.propagatedType;
+    if (leftStaticType.name == _dynamicClassName &&
```

Comment:
or leftStaticType.isDynamic

Also, I suspect that you can simplify the logic by using `bestType` rather than either `staticType` or `propagatedType`.